blackduck/Client.py

```diff
@@ -10,9 +10,55 @@
 from .Utils import find_field, safe_get
 from .Authentication import BearerAuth
 import logging
-import requests
+import os
+import requests.packages.urllib3
+from requests.packages.urllib3.util.retry import Retry
+from requests.adapters import HTTPAdapter
+from urllib.parse import urljoin
+
 logger = logging.getLogger(__name__)
 
+
+class HubSession(requests.Session):
+    """Hold base_url, timeout, retries, and provide sensible defaults"""
+
+    def __init__(self, base_url, timeout, retries, verify):
+        super().__init__()
+        self.base_url = base_url
+        self._timeout = float(timeout)  # timeout is not a member of requests.Session
+        self.verify = verify
+
+        # use sane defaults to handle unreliable networks
+        """HTTP response status codes:
+                429 = Too Many Requests
+                500 = Internal Server Error
+                502 = Bad Gateway
+                503 = Service Unavailable
+                504 = Gateway Timeout
+        """
+        retry_strategy = Retry(
+            total=int(retries),
+            backoff_factor=2,  # exponential retry 1, 2, 4, 8, 16 sec ...
+            status_forcelist=[429, 500, 502, 503, 504],
+            method_whitelist=['GET']
+        )
+
+        adapter = HTTPAdapter(max_retries=retry_strategy)
+        self.mount("https://", adapter)
+        self.mount("http://", adapter)
+        logging.info("Using a session with a %s second timeout and up to %s retries per request", timeout, retries)
+
+        self.proxies.update({
+            'http': os.environ.get('http_proxy', ''),
+            'https': os.environ.get('https_proxy', '')
+        })
+
+    def request(self, method, url, **kwargs):
+        kwargs['timeout'] = self._timeout
+        url = urljoin(self.base_url, url)
```

Comment:
Actually, would this work with linked resources i.e. `urljoin('app.blackduck.com/', 'app.blackduck.com/projects/id/versions/id')`

Reply:
No it does not, result is "app.blackduck.com/app.blackduck.com/projects/id/versions/id". I'll look into the resources a bit more closely and determine a way to traverse them.

Reply:
If the doubleslashes are in the URLs, urljoin detects the 'netloc' and joins them correctly. E.g.:
urljoin("https://app.blackduck.com", "https://app.blackduck.com/projects/id/versions/id") AND
urljoin("https://app.blackduck.com", "//app.blackduck.com/projects/id/versions/id")
both give "https://app.blackduck.com/projects/id/versions/id".

Linked resources in Hub's REST API have the full URL. Urljoin'ing this with the base_url is giving the desired result.